src/irx/builders/llvmliteir.py

```diff
@@ -264,20 +264,37 @@ def promote_operands(
                 rhs = self._llvm.ir_builder.sext(rhs, lhs.type, "promote_rhs")
             return lhs, rhs
 
-        # ranking dictionary for floating point types
-        fp_types_order = {"float": 1, "double": 2}
-
-        lhs_str = str(lhs.type)
-        rhs_str = str(rhs.type)
-
-        # perform floating-point extension
-        if lhs_str in fp_types_order and rhs_str in fp_types_order:
-            if fp_types_order[lhs_str] < fp_types_order[rhs_str]:
+        def fp_rank(t: ir.Type) -> int:
+            # ranks: half < float < double
+            s = str(t)
+            if "half" in s or isinstance(t, ir.HalfType):
+                return 1
+            if "float" in s and "double" not in s:
+                return 2
+            if "double" in s or isinstance(t, ir.DoubleType):
+                return 3
+            return 0
+
+        if fp_rank(lhs.type) > 0 and fp_rank(rhs.type) > 0:
```

Comment:
instead of running multiple times fp_rank for lhs or rhs ... maybe you could create a variable and store that just once:

```python
lhs_fp_rank = fp_rank(lhs.type)
rhs_fp_rank = fp_rank(rhs.type)
```